### DISCUSSION TITLE
Volatile and Cross-Thread Variable Changes

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/runtime#78346] [Q&A]
What is the correct way to ensure that variables changed on one thread are observed on another?

Also, why is it that a ```long``` cannot be declared as volatile but you can ```Volatile.Read/Write``` longs?  I expected that the API would be consistent with the keyword.

**Comment by CyrusNajmabadi** [comment]
Generally speaking, i view the correct way as using a normal 'lock'.  If there are exceptional circumstances that do not allow for that, then the answer will depend on those circumstances.

**Comment by alexrp** [comment]
> Also, why is it that a `long` cannot be declared as volatile but you can `Volatile.Read/Write` longs? I expected that the API would be consistent with the keyword.

Unfortunate historical reasons. The compiler does not use that API at all.

C#'s `volatile` gives you acquire/release semantics + atomicity. But at the CIL level, it is emitted as the `volatile.` prefix opcode (giving the acquire/release semantics only) + a regular load/store. In CIL, regular loads/stores are only atomic for a given type `T` if `sizeof(T) <= sizeof(void*)`. That means that a C#-level `volatile long` cannot be guaranteed to be atomic on 32-bit machines. So the language forbids it to avoid the footgun.

AFAIK, there's no technical reason why the compiler has to continue to emit `volatile` in this way. For newer TFMs (i.e. anything .NET Standard 1.0+) it could theoretically be changed to emit calls to the `Volatile` class, and this restriction could be lifted. But I don't think there's much appetite for doing this, partly because of compatibility concerns, and partly because adding `volatile` to the language has been widely regarded as a mistake inherited from C/C++.

**Comment by TonyValenti** [comment]
Thanks so much for the answers here. 

With my code, I want all threads to always be able to read the current version of the variables (ideally, without a lock).

When updating the variables, I do lock.

Will a volatile variable that has a Lock around all writes guarantee that all threads always see the current value?

**Comment by NN---** [comment]
This question is not new :)

https://stackoverflow.com/questions/4727068/why-not-volatile-on-system-double-and-system-long
